components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt

```diff
@@ -78,68 +71,27 @@ class GeckoLoginStorageDelegate(
 
     @Synchronized
     override fun onLoginSave(login: Login) {
+        val validationDelegate = DefaultLoginValidationDelegate(storage = loginStorage)
         scope.launch {
             val existingLogin = login.guid?.let { loginStorage.value.get(it) }
-
-            when (getPersistenceOperation(login, existingLogin)) {
-                Operation.UPDATE -> {
-                    existingLogin?.let { loginStorage.value.update(it.mergeWithLogin(login)) }
+                ?: loginStorage.value.getStoredLoginByImportantFields(login)
```

Comment:
This is quite inefficient - inside of `getNewLoginPersistenceAsync` you'll be running both of these "getters"!
It's also very complex to actually reason about what's going on here.

Reply:
Okay this was a very valid criticism :)
I changed this a bit so that the `CanBeUpdated` `Result` needs to pass which login we found back to the caller. Should help with being more explicit about what is going on here too.